### PR TITLE
feature: new parallel e2e tf apply destroy workflows

### PR DESCRIPTION
**Pull request by Zvikan**
### What does this PR do?

- Adds two new GH workflows:
  - full **parallel** e2e for configured examples (TF plan->apply->destroy)
  - destroy only workflow e2e parallel for configured examples

- Using `-target` module instead of single apply/destroy for correct deployment order. 
- Template E2E deploy folder that's being used by the e2e workflow (reduce the need of code duplication per example to be supported by the parallel e2e).
- Adding variables file for each example to allow E2E workflow to create multiple clusters at the same time without failing due to similar cluster name (tenant/env/zone).

### Motivation

- We need a way to be able to test all our examples in automated way, to begin with, run full e2e of TF plan, apply and destroy.
- Support examples that require specific deployment order (e.g. fully private cluster).
- Support new customized deploy examples (e.g. examples deployed via `deploy` folder).
- Solution that can be scaled fairly easily and with minimize amount of work to on-board new examples in the future.

### More

- [ ] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/ssp-eks-add-ons) repo (if applicable)
- [ ] Yes, I have updated the [docs](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: .github/workflows/e2e-parallel-full.yml [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - example_path: examples/analytics/emr-on-eks
+            tenant_name: "emr-eks"
+          - example_path: examples/analytics/spark-k8s-operator
+            tenant_name: "spark"
+          # Disabling as there's work in progress to fix issues in https://github.com/aws-samples/ssp-eks-add-ons
+          # - example_path: examples/argocd
```
Comment (kcoleman731) [inline .github/workflows/e2e-parallel-full.yml:31]: lets get rid of commented config

File: deploy/e2e/gh-e2e-template/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Usage
```
Comment (kcoleman731) [inline deploy/e2e/gh-e2e-template/README.md:1]: can you remove if we don't have any comments?

**Review by kcoleman731 — COMMENTED** [review]
Amazing. Comments for ya.

File: .github/workflows/e2e-parallel-full.yml [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        continue-on-error: false
+
+      - name: Setup Deployment Order
+        id: setup-deployment-order
+        run: |
+          export MATRIX_DEPLOYMENT_ORDER=${{join(matrix.deployment_order,',') }}
+
+          if [ -z "$MATRIX_DEPLOYMENT_ORDER" ]
+          then
+                DEPLOYMENT_ORDER=${{env.DEFAULT_DEPLOY_ORDER}}
```
Comment (askulkarni2) [inline .github/workflows/e2e-parallel-full.yml:91]: nit: indentation is different

File: deploy/e2e/gh-e2e-template/README.md (hunk 2 above)
Comment (Zvikan) [reply deploy/e2e/gh-e2e-template/README.md:1]: Eh.....I'm not sure I can, the github action flow (terraform-docs) added it

**Review by askulkarni2 — APPROVED** [review]
LGTM!

File: .github/workflows/e2e-parallel-full.yml [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+          if [[ ${{ matrix.example_path }} == deploy/* ]]
+          then
+            echo "Skipping pre-setup for ${{ matrix.example_path }}"
+            cp -R ${{ matrix.example_path }}/* deploy/e2e/gh-e2e-test/
+          else 
+            echo "Running pre-setup for ${{ matrix.example_path }}"
+            mkdir -p deploy/e2e/gh-e2e-test
+            cp -R deploy/e2e/gh-e2e-template/* deploy/e2e/gh-e2e-test/
+            sed -i "s!REPLACE_ME!${{ matrix.tenant_name }}!g" deploy/e2e/gh-e2e-test/base.tfvars
+            sed -i "s!TF_STATE_PATH!${{ matrix.example_path }}!g" deploy/e2e/gh-e2e-test/backend.conf
```
Comment (vara-bonthu) [inline .github/workflows/e2e-parallel-full.yml:81]: `backend.conf` file is missing in the PR. Is this file meanto be part of `deploy/e2e/gh-e2e-template/`?

File: .github/workflows/e2e-parallel-destroy.yml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            echo "Running pre-setup for ${{ matrix.example_path }}"
+            mkdir -p deploy/e2e/gh-e2e-test
+            cp -R deploy/e2e/gh-e2e-template/* deploy/e2e/gh-e2e-test/
+            sed -i "s!REPLACE_ME!${{ matrix.tenant_name }}!g" deploy/e2e/gh-e2e-test/base.tfvars
+            sed -i "s!TF_STATE_PATH!${{ matrix.example_path }}!g" deploy/e2e/gh-e2e-test/backend.conf
+            sed -i "s!EXAMPLE_PATH!${{ matrix.example_path }}!g" deploy/e2e/gh-e2e-test/main.tf
+            cat deploy/e2e/gh-e2e-test/base.tfvars
+            cat deploy/e2e/gh-e2e-test/backend.conf
+            cat deploy/e2e/gh-e2e-test/main.tf
+            tree -a
```
Comment (vara-bonthu) [inline .github/workflows/e2e-parallel-destroy.yml:84]: Do you need these lines? Is this added for debugging?

File: .github/workflows/e2e-parallel-destroy.yml [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+          if [[ ${{ matrix.example_path }} == deploy/* ]]
+          then
+            echo "Skipping pre-setup for ${{ matrix.example_path }}"
+            cp -R ${{ matrix.example_path }}/* deploy/e2e/gh-e2e-test/
+          else 
+            echo "Running pre-setup for ${{ matrix.example_path }}"
+            mkdir -p deploy/e2e/gh-e2e-test
+            cp -R deploy/e2e/gh-e2e-template/* deploy/e2e/gh-e2e-test/
+            sed -i "s!REPLACE_ME!${{ matrix.tenant_name }}!g" deploy/e2e/gh-e2e-test/base.tfvars
+            sed -i "s!TF_STATE_PATH!${{ matrix.example_path }}!g" deploy/e2e/gh-e2e-test/backend.conf
```
Comment (vara-bonthu) [inline .github/workflows/e2e-parallel-destroy.yml:79]: See other comments on `backend.conf`

File: .github/workflows/e2e-parallel-full.yml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            sed -i "s!EXAMPLE_PATH!${{ matrix.example_path }}!g" deploy/e2e/gh-e2e-test/main.tf
+            cat deploy/e2e/gh-e2e-test/base.tfvars
+            cat deploy/e2e/gh-e2e-test/backend.conf
+            cat deploy/e2e/gh-e2e-test/main.tf
+            tree -a
```
Comment (vara-bonthu) [inline .github/workflows/e2e-parallel-full.yml:86]: Same as the other comment. See above

File: .github/workflows/e2e-parallel-full.yml (hunk 4 above)
Comment (Zvikan) [reply .github/workflows/e2e-parallel-full.yml:81]: Good catch! fixing.

File: .github/workflows/e2e-parallel-destroy.yml (hunk 5 above)
Comment (Zvikan) [reply .github/workflows/e2e-parallel-destroy.yml:84]: Can probably be cleaned up, mostly useful to see that the sed commands were done correctly.

**Review by kcoleman731 — COMMENTED** [review]
LGTM

**Pull request merged**